Unbreak mailctl

aidalgol opened the pull request:
## Description of changes
A recent hackage auto-update indirectly broke `mailctl`.  This fixes the broken dependency, and adds it to the package list in `pkgs/top-level/release-haskell.nix` to prevent this from happening again.

Thanks to @maralorn for assisting with this fix.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

aidalgol committed:
mailctl: add to release-haskell.nix

This prevents Hackage updates from silently breaking this package.

aidalgol committed:
haskellPackages.twain: build against supported version of http2

aidalgol committed:
mailctl: fix broken transitive dependency

maralorn commented on pkgs/development/haskell-modules/configuration-hackage2nix/main.yaml:
```diff
@@ -112,6 +112,7 @@ extra-packages:
   - hspec-discover < 2.8                # 2022-04-07: Needed for tasty-hspec 1.1.6
   - hspec-meta < 2.8                    # 2022-12-07: Needed for elmPackages.elm / hspec-discover
   - hspec-golden == 0.1.*               # 2022-04-07: Needed for elm-format
+  - http2 < 3.3                         # 2023-08-24: Needed for twain
```
Can you add a comment which twain version needs the override and a link to a pr or issue upstream.

aidalgol replied:
From looking at twain's git history, it has depended upon http2 since version 2.0.0.0 and has had the upper bounds of http2 3.x ever since.  There is no upstream issue for migrating it to the current version of http2.
https://github.com/alexmingoia/twain/commit/2ef20f5df66d1cddb42e8b672d92267c556fdcd5#diff-99c6c7a2194a963526fd5d0f57663534b8b754f09e8ff44717397578ee585a9e

maralorn replied:
Normally, in this case, we open an issue (or PR depending on time and motivation) upstream, where we friendly point out, that there is a newer release of the dependency.

aidalgol replied:
I just created an issue, as I barely know Haskell, so I could not do the work myself.

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -314,6 +314,8 @@ self: super: {
   # defined as >= 3.1  && < 3.2, can be removed once pandoc gets bumped by Stackage.
   patat = super.patat.override { pandoc = self.pandoc_3_1_6; };
 
+  twain = super.twain.override { http2 = self.http2_3_0_3; };
```
Can you add a comment that this override is repeated in all-packages.nix so that we don’t forget to remove it there as well?

aidalgol replied:
Done.

maralorn commented on pkgs/top-level/all-packages.nix:
```diff
@@ -2654,7 +2654,11 @@ with pkgs;
   # Does not build with default Haskell version because upstream uses a newer Cabal version.
-  mailctl = haskell.packages.ghc94.callPackage ../tools/networking/mailctl { };
+  mailctl = (haskell.packages.ghc94.callPackage ../tools/networking/mailctl {}).overrideScope (final: prev: {
+    # Dependency twain requires an older version of http2, and we cannot safely
```
I would remove the "safely".

maralorn replied:
Oh, and unless there is a good reason prefer using haskellPackages over haskell.packages.ghc94 so that this override does not go stale.

aidalgol replied:
Done and done.

maralorn reviewed (approved):
Great, thx! I have added a few remarks to make the comments even more informative.

maralorn commented:
Oh, and please change the PR to merge into to haskell-updates.

aidalgol committed:
mailctl: use haskellPackages

mailctl now builds against current haskellPackages, so the override
is unneeded.

maralorn commented:
Well done! In the meantime twean got marked broken on our branch. I will quickly fix that.
